sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Redistribute.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import java.util.concurrent.ThreadLocalRandom;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.util.IdentityWindowFn;
+import org.apache.beam.sdk.util.Reshuffle;
+import org.apache.beam.sdk.util.ReshuffleTrigger;
+import org.apache.beam.sdk.util.ValueWithRecordId;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+
+/**
+ * {@link PTransform Transforms} that return a {@link PCollection} equivalent to their input, but
+ * which is distributed differently, in order to provide additional operational properties. See
+ * {@link #byKey} and {@link #arbitrarily}.
+ */
+public class Redistribute {
+  /**
+   * Redistributes the collection by key, providing some of the side effects of a {@link
+   * GroupByKey}, in particular: prevents fusion of the surrounding transforms, checkpointing and
+   * deduplication by id (see {@link ValueWithRecordId}), and ensures that elements with the same
+   * key are processed by the same worker.
+   *
+   * @param <K> The type of key being reshuffled on.
+   * @param <V> The type of value being reshuffled.
+   */
+  public static <K, V> PTransform<PCollection<KV<K, V>>, PCollection<KV<K, V>>> byKey() {
+    return new ByKey<>();
+  }
+
+  /**
+   * Redistributes the collection in an arbitrary fashion, preventing fusion of the surrounding
+   * transforms.
+   *
+   * @param <T> Type of the value being redistributed.
+   */
+  public static <T> PTransform<PCollection<T>, PCollection<T>> arbitrarily() {
+    return new Arbitrarily<>();
+  }
+
+  /** Implementation of {@link #arbitrarily}. */
+  private static class Arbitrarily<T> extends PTransform<PCollection<T>, PCollection<T>> {
+    @Override
+    public PCollection<T> apply(PCollection<T> input) {
+      return input
+          .apply(ParDo.of(new PairWithBundleKey<T>()))
+          .apply(Reshuffle.<Integer, T>of())
```

Comment:
In the email thread you said it "mostly subsumes" `Reshuffle`, but it is defined in terms of it?